cpp/src/arrow/compute/kernels/scalar_cast_extension.cc

```diff
@@ -0,0 +1,74 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// Implementation of casting to extension types
+#include "arrow/compute/kernels/common.h"
+#include "arrow/compute/kernels/scalar_cast_internal.h"
+#include "arrow/scalar.h"
+
+namespace arrow {
+namespace compute {
+namespace internal {
+
+namespace {
+Status CastToExtension(KernelContext* ctx, const ExecSpan& batch, ExecResult* out) {
+  const CastOptions& options = checked_cast<const CastState*>(ctx->state())->options;
+  auto out_ty = static_cast<const ExtensionType&>(*options.to_type.type).storage_type();
+
+  DCHECK(batch[0].is_array());
+  std::shared_ptr<Array> array = batch[0].array.ToArray();
+
+  // Try to prevent user errors by preventing casting between extensions w/
+  // different storage types. Provide a tip on how to accomplish same outcome.
+  if (array->type()->id() == Type::EXTENSION) {
+    const auto& ext_arr = checked_cast<const ExtensionArray&>(*array);
+    if (out_ty->id() != ext_arr.extension_type()->storage_id()) {
```

Comment:
My interpretation was to (for now) disallow any cast unless the input array has exactly a type equal to the target storage type (in which case basically no actual cast is needed, only the creation of the extension array ..)

So something like `if (!array->type()->Equals(out_ty)) { ...`

Reply:
Ah okay.. then it's updated, and also does not cast unless it's needed.

Reply:
For future reference, Miles and I chatted about this and were thinking to keep the above, and thus only disallow Extension->Extension casts for now in this PR, while still allow Any->Extension casts based on the storage type casting rules (so if Any can be cast to storage type, then it can also be cast to the Extension type).

This would make it consistent with the existing Extension->Any cast, which also just defers to the casting rules of the storage type at the moment (and does not limit to strictly casting Extension only to the exact storage type). And I think this is also what others were OK with (based on the inline discussion in https://github.com/apache/arrow/pull/14106#discussion_r977663050)

Personally I still think it's a bit inconsistent to allow casting any unrelated type to ExtensionType, but not allow an unrelated extension array to ExtensionType. But we can see to allow registering such casts in the follow-up https://issues.apache.org/jira/browse/ARROW-17890

Reply:
Implicit conversions always have to end up inconsistent in practice, so this is not a problem.

For example, in Python you can convert a `dict` to `bool`, a `bool` to `float`, but not a `dict` to `float`.

Reply:
Yes, but you can use that analogy for the "any -> extension<storage>" casts as well (for the case where "any" is not equal to "storage"), that the user needs to do "any -> storage -> extension<storage>" explicitly instead .

Reply:
> Yes, but you can use that analogy for the "any -> extension" casts as well (for the case where "any" is not equal to "storage"), that the user needs to do "any -> storage -> extension" explicitly instead .

Well, I wouldn't mind this :-)